File: clojure/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn question? [s]
  (= \? (last s)))

(defn nothing? [s]
  (re-matches #"^\s*$" s))

(defn yelling? [s]
  (and (= s (str/upper-case s))
       (re-find #"[A-Z]{1,}" s)))

(defn response-for [s]
  (let [s (str/trim s)]
    (cond
      (and (yelling? s) (question? s)) "Calm down, I know what I'm doing!"
      (yelling? s) "Whoa, chill out!"
      (question? s) "Sure."
      (nothing? s) "Fine. Be that way!"
      :else "Whatever.")))

File: rust/bob.rs
fn is_all_caps(message: &str) -> bool {
    has_letters(message) && message.to_uppercase() == message
}

fn has_letters(message: &str) -> bool {
    return message.chars().filter(|x| x.is_alphabetic()).count() > 0;
}

pub fn reply(message: &str) -> &str {
    let m = message.trim();
    let mut the_reply: &str = "Whatever.";
    if is_all_caps(m) && m.ends_with("?") {
        the_reply = "Calm down, I know what I'm doing!";
    } else if m.ends_with("?") {
        the_reply = "Sure.";
    } else if is_all_caps(m) {
        the_reply = "Whoa, chill out!";
    } else if m.trim().is_empty() {
        the_reply = "Fine. Be that way!"
    }
    the_reply
}
